### PR TITLE
Add oauth2 http proxy client

### PR DESCRIPTION
**Pull request by ByronHsu**
# TL;DR

Add http_proxy to oauth2 client because in on-prem cluster use cases, the client might not have access to the external network

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

https://github.com/flyteorg/flyte/issues/3620

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: clients/go/admin/auth_interceptor.go [hunk 1]
```diff
@@ -50,6 +52,21 @@ func shouldAttemptToAuthenticate(errorCode codes.Code) bool {
...
+
+	if len(cfg.HTTPProxyURL.String()) > 0 {
+		// create a transport that uses the proxy
+		transport := &http.Transport{
+			Proxy: http.ProxyURL(&cfg.HTTPProxyURL.URL),
+		}
+		httpClient.Transport = transport
+	}
+
+	return context.WithValue(ctx, oauth2.HTTPClient, httpClient)
```
Comment (wild-endeavor) [inline clients/go/admin/auth_interceptor.go:67]: the client in the context will automatically get picked up?

File: clients/go/admin/auth_interceptor.go [hunk 2]
```diff
@@ -62,6 +79,8 @@ func shouldAttemptToAuthenticate(errorCode codes.Code) bool {
 // be able to find and acquire a valid AccessToken to annotate the request with.
 func NewAuthInterceptor(cfg *Config, tokenCache cache.TokenCache, credentialsFuture *PerRPCCredentialsFuture) grpc.UnaryClientInterceptor {
 	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		ctx = setHTTPClientContext(ctx, cfg)
+
 		err := invoker(ctx, method, req, reply, cc, opts...)
```
Comment (wild-endeavor) [inline clients/go/admin/auth_interceptor.go:84]: like this invoker will automatically use the client created in `setHTTPClientContext`?

**Review by wild-endeavor — APPROVED** [review]
just questions.

File: clients/go/admin/auth_interceptor.go [hunk 3]
```diff
@@ -50,6 +52,21 @@ func shouldAttemptToAuthenticate(errorCode codes.Code) bool {
 	return errorCode == codes.Unauthenticated
 }
 
+// Set up http client used in oauth2
+func setHTTPClientContext(ctx context.Context, cfg *Config) context.Context {
+	httpClient := &http.Client{}
```
Comment (wild-endeavor) [inline clients/go/admin/auth_interceptor.go:57]: and there's no issue with the overhead of creating these clients right?  if not specified the default client in the http lib gets used it looks like.

File: clients/go/admin/auth_interceptor.go (hunk 1 above)
Comment (ByronHsu) [reply clients/go/admin/auth_interceptor.go:67]: yes

File: clients/go/admin/auth_interceptor.go (hunk 2 above)
Comment (ByronHsu) [reply clients/go/admin/auth_interceptor.go:84]: yes

File: clients/go/admin/auth_interceptor.go (hunk 3 above)
Comment (ByronHsu) [reply clients/go/admin/auth_interceptor.go:57]: if not specified it will just be a `&http.Client{}`

**Review by wild-endeavor — APPROVED** [review]

**Comment by pingsutw** [comment]
@ByronHsu would you mind closing this pr, and open a new one again. CI stuck for some reason...

**Review by pingsutw — APPROVED** [review]
